tests/test_logging.py

```diff
@@ -38,8 +40,45 @@ def test_getLevelName(self):
         self.assertEqual(getLevelName('FATAL'), logging.ERROR)
         self.assertEqual(getLevelName('OFF'), logging.CRITICAL)
 
-    def tearDown(self):
-        initLogging()
+    def test_logging_non_duplicate(self):
+        """
+        Verify that child loggers don't propagate a log message they handle
+        """
+
+        root_logger = logging.getLogger('')
+        self.assertTrue(root_logger.handlers, 'root logger has at least 1 handler')
+
+        parent_capture = FIFOIO(256)
+        parent_handler = logging.StreamHandler(parent_capture)
+        parent_handler.setFormatter(logging.Formatter(fmt=LOG_FORMAT, datefmt=LOG_DATEFMT))
+        parent_logger = getLogger('foo')
+        self.assertTrue(parent_logger.propagate, 'no handler on logger => propagate')
+        parent_logger.addHandler(parent_handler)
+        parent_logger = getLogger('foo')
+        self.assertFalse(parent_logger.propagate, 'should not propagate because StreamHandler has been attached')
+
+        child_capture = FIFOIO(256)
+        child_handler = logging.StreamHandler(child_capture)
+        child_handler.setFormatter(logging.Formatter(fmt=LOG_FORMAT, datefmt=LOG_DATEFMT))
+        child_logger = getLogger('foo.bar')
+        self.assertTrue(child_logger.propagate, 'no handler on logger => propagate')
+        child_logger.addHandler(child_handler)
+        child_logger = getLogger('foo.bar')
+        self.assertFalse(child_logger.propagate, 'should not propagate because StreamHandler has been attached')
+
+        child_logger.error('test')
+
+        parent_output = parent_capture.getvalue()
+        parent_capture.close()
+        child_output = child_capture.getvalue()
+        child_capture.close()
+        # print('parent', parent_output)
+        print('child', child_output)
+
+        # self.assertEqual(parent_output, '', 'parent logger should not receive msg')
```

Comment:
Why is this commented out?

Another configuration worth testing: before adding a handler to `child_logger`, send a message and validate that `parent_output` received it. (So, combining both, you could assert the same condition for both outputs – which now validates that `parent_handler` received only the first message.)

Also, how about log levels? You could set `child_handler` to `DEBUG` and send a debug message to `child_logger`, both before and after you attach `child_handler`. Then you assert that `child_output` received it only once. (This validates that `parent_handler` ignored this level.)

Reply:
> Why is this commented out?

Forgot to comment back in before commit, fixed.

> Another configuration worth testing: before adding a handler to child_logger, send a message and validate that parent_output received it. (So, combining both, you could assert the same condition for both outputs – which now validates that parent_handler received only the first message.)

> Also, how about log levels? You could set child_handler to DEBUG and send a debug message to child_logger, both before and after you attach child_handler. Then you assert that child_output received it only once. (This validates that parent_handler ignored this level.)

See https://github.com/OCR-D/core/pull/462/commits/43d905d4834c6ec90b1ca332da51fdd8953eff01 Sends two error and two debug messages to child_logger. parent handler receives only the first error (because at that point no handler on child_logger and first debug message below parent_logger's level) and child_handler receives only second error and second debug.

Reply:
Excellent!